Lib9c/Action/HackAndSlash4.cs

```diff
@@ -0,0 +1,281 @@
+
+            sw.Restart();
+
+            if (avatarState.RankingMapAddress != RankingMapAddress)
+            {
+                throw new InvalidAddressException("Invalid ranking map address");
+            }
+
+            // worldId와 stageId가 유효한지 확인합니다.
+            var worldSheet = states.GetSheet<WorldSheet>();
+
+            if (!worldSheet.TryGetValue(worldId, out var worldRow, false))
+            {
+                throw new SheetRowNotFoundException(nameof(WorldSheet), worldId);
+            }
+
+            if (stageId < worldRow.StageBegin ||
+                stageId > worldRow.StageEnd)
+            {
+                throw new SheetRowColumnException(
+                    $"{worldId} world is not contains {worldRow.Id} stage: " +
+                    $"{worldRow.StageBegin}-{worldRow.StageEnd}");
+            }
+
+            var stageSheet = states.GetSheet<StageSheet>();
+            if (!stageSheet.TryGetValue(stageId, out var stageRow))
+            {
+                throw new SheetRowNotFoundException(nameof(StageSheet), stageId);
+            }
+
+            var worldInformation = avatarState.worldInformation;
+            if (!worldInformation.TryGetWorld(worldId, out var world))
+            {
+                // NOTE: Add new World from WorldSheet
+                worldInformation.AddAndUnlockNewWorld(worldRow, ctx.BlockIndex, worldSheet);
+            }
+
+            if (!world.IsUnlocked)
+            {
+                throw new InvalidWorldException($"{worldId} is locked.");
+            }
+
+            if (world.StageBegin != worldRow.StageBegin ||
+                world.StageEnd != worldRow.StageEnd)
+            {
+                worldInformation.UpdateWorld(worldRow);
+            }
+
+            if (world.IsStageCleared && stageId > world.StageClearedId + 1 ||
+                !world.IsStageCleared && stageId != world.StageBegin)
+            {
+                throw new InvalidStageException(
+                    $"Aborted as the stage ({worldId}/{stageId}) is not cleared; " +
+                    $"cleared stage: {world.StageClearedId}"
+                );
+            }
+
+            avatarState.ValidateEquipments(equipments, context.BlockIndex);
+            avatarState.ValidateConsumable(foods, context.BlockIndex);
+            avatarState.ValidateCostume(new HashSet<int>(costumes));
+
+            var costumeStatSheet = states.GetSheet<CostumeStatSheet>();
+
+            sw.Restart();
+            if (avatarState.actionPoint < stageRow.CostAP)
+            {
+                throw new NotEnoughActionPointException(
+                    $"Aborted due to insufficient action point: " +
+                    $"{avatarState.actionPoint} < {stageRow.CostAP}"
+                );
+            }
+
+            avatarState.actionPoint -= stageRow.CostAP;
+
+            avatarState.EquipCostumes(new HashSet<int>(costumes));
+
+            avatarState.EquipEquipments(equipments);
```

Comment:
```suggestion
            var itemsToEquip = new List<Guid>(costumes).AddRange(equipments);
            avatarState.EquipItems(itemsToEquip);
```

https://github.com/planetarium/lib9c/blob/6eb98d42ccd4cc7bff86688af6bbb74e01c84d19/Lib9c/Model/State/AvatarState.cs#L522